### ISSUE TITLE
Precision drop?

### ISSUE THREAD

**Issue by vivoutlaw** [varun-suresh/Clustering#1]
Hi Varun, 

were you able to resolve, why your implementation has such a huge drop in precision? Thanks and looking forward to your reply.

best, Vivek

**Comment by varun-suresh** [comment]
Hi Vivek,
I haven't had the time to revisit it, I will take a look soon and see if I can figure it out.

**Comment by vivoutlaw** [comment]
ok, thanks

**Comment by gmy001** [comment]
Hi Varun,
The code looks great but I can't get the f1=0.52 result as you mentioned in README. 
Here is what I did:
1.Download the face vectors from the link in README (I chose LightenedCNN_C_lfw.mat).
2.Run demo.py, I removed the '#' of line 109-118 to perform evaluation.
With the default threshold I got an f1=0.26, I tried changing thresholds but f1=0.3 was the best result I got. What is wrong with my steps? Do you have any idea? Thanks a lot!

**Comment by varun-suresh** [comment]
Hey,
If I remember correctly, I got the result using LightenedCNN_A_lfw.mat. I was testing with LightenedCNN_C_lfw.mat yesterday and I got the same results you did. 
I'm trying to visualize the results and see if I can figure out what is going on. I will post here once I have an insight into it.

**Comment by gmy001** [comment]
Thanks! I will try LightenedCNN_A_lfw.mat and other features as well :)

**Comment by varun-suresh** [comment]
I have updated the repository. If any of you have time to generate the face vectors using VGG or OpenFace, it would be really interesting to check. 

Thanks!

**Comment by gmy001** [comment]
Hi Varun,
    Thanks for the update! These days I tried using [FaceNet](https://github.com/davidsandberg/facenet) feature, which utilizes L2 dist as metric. Still I got similar result after running demo.py: the best f1 score was 0.3(thresh=2.1). I visualized the 700+ clusters, there was a huge cluster containing more than 12K faces and all rest clusters contains only 1~4 faces. I looked in each cluster and they are just not very accurate. [Here ](https://drive.google.com/drive/folders/15eR6bbpuc0MWZ4q-9agoN6G5RQZg-qP0) is the feature mat file, I made a pdist2 check and it should be fine.
    Despite FaceNet feature, I also performed length-1 normalization on LightenedCNN_C_lfw.mat feature, which should make cosine distance and L2 distance equivalent, but still the result was not good.
    I think something goes wrong with that paper but I don't have any clue:(

Mengyue

**Comment by gmy001** [comment]
OK, OK, after another detailed code check I finally found the problems:
1.   Function 'create_labels_lookup' in demo.py has a line labels_lookup[idx] = int(label[0][0]), this will capture only the first digit of the label, so the loaded gt labels are seriously wrong, which causes the evaluation problem, the correct line should be labels_lookup[idx] = int(label[0][:]). Fixing this issue will get f1=0.75, but still precision drops drastically when changing thresh from 1.9 to 2.0. 

2.   I checked the paper again and found that in function 'calculate_symmetric_dist_row'(clustering.py), the code counts the shared neighbor number, and divide by max(Oi,Oj), where Oj can be K+1, i.e sample i may not be in sample j's nearest neighbor list. This is somehow not correct since the paper said it only calculate Dij for which i and j are **both** in each other's neighbor list. So I made a change which set dij = 9999.0 if j is not in i's neighbor list. Another issue is that the paper only counts shared neighbor of i and j to from index 0 to the index of Oj and Oi, not the full K list, this is also fixed.
    After the above changes, f1_score moves smoothly between different thresholds, and maximum f1_score(by using LightenedCNN_C_lfw) was 0.88~0.9, matches the result in the paper.
    I have put the fixed version [here](https://github.com/gmy001/Clustering). Again thanks for the excellent implementation:)
